### ISSUE TITLE
Bug with non-breaking spaces in 0.3.0?

### ISSUE THREAD

**Issue by ulrichsg** [soundasleep/html2text#29]
Hi,

I've found a very weird case where html2text returned broken output after upgrading to 0.3.0, and I was able to narrow it down to this line:

`$html = str_replace("\xa0", " ", $html);`

I examined the contents of `$html` in a hex editor immediately before and after the line and found this diff. The original source snippet reads "für Ihre ", with the spaces being nbsp's, and is UTF-8 encoded.

Before: `66 c3 bc 72 c2 a0 49 68 72 65 c2 a0`
After: `66 c3 83 c2 bc 72 c3 82 20 49 68 72 65 c3 82 20`

So apparently the nbsp's (`c2 a0`) have been transformed into `c3 82 20`, which looks like a regular space (`20`) but with some gibberish in front of it. Also, the multi-byte character 'ü' (`c3 bc`) is now `c3 83 c2 bc`, which is also nonsensical.

I've downgraded to 0.2.3 and all is fine now, but I'd like to let you know in case you'd like to look into this.

**Comment by soundasleep** [comment]
This definitely sounds like sadness, some weird interaction between UTF-8 and str_replace. Maybe there needs to be some utf8_encode()/utf8_decode(), maybe there needs to be some iconv/mbstring, maybe html2text simply can't support this... What was the hex output of html2text 0.2.3 on the same input?

Can you create a pull request with a new test (in tests/), both with the before and expected after text encoded as UTF-8, and we can go from there?

**Comment by ulrichsg** [comment]
Sure thing :)

**Comment by jigante** [comment]
I have the same problem. All the content that was converted correctly before the upgrade from 0.2.3 to 0.3.0 now brings question marks characters, not recognized as UTF8, breaking the output.

Removing the line

`$html = str_replace("\xa0", " ", $html);`

from "Html2Text" class resolves the problem

**Comment by soundasleep** [comment]
Can you all check out the new `non-breaking-space` branch and see if this resolves the issue for your tests? (PHP + character encodings = much sadness!)

If so, I'll merge into `master` and release 0.3.1 right away. If not, can you please contribute more pull requests off this branch, with additional tests?

Thank you @ulrichsg for the test contribution!

**Comment by jigante** [comment]
The new commit in the `non-breaking-space` branch solves the problem for my use cases

**Comment by soundasleep** [comment]
Resolved with PR #32 and released as 0.3.1
